runtime/common/ArgumentConversion.cpp

```diff
@@ -146,9 +379,136 @@ static Value genConstant(OpBuilder &builder, const cudaq::state *v,
+  //   %1 = quake.extract_ref %arg0[0] : (!quake.veq<2>) -> !quake.ref
+  //   quake.x %1 : (f64, !quake.ref) -> ()
+  //   return
+  // }
+  // ```
+  // clang-format on
+  //
+  // 2. Replace the `quake.get_state` ops with calls to the generated functions
+  //    synthesized with the arguments used to create the state:
+  //
+  // After ReplaceStateWithKernel pass:
+  //
+  // clang-format off
+  // ```
+  // func.func @caller() {
+  //   %1 = call 2callee.num_qubits_0() : () -> i64
+  //   %2 = quake.alloca !quake.veq<?>[%1 : i64]
+  //   %3 = call @callee.init_0(%2): (!quake.veq<?>) -> !quake.veq<?>
+  // }
+  //
+  // func.func private @callee.num_qubits_0() -> i64 {
+  //   %cst = arith.constant 2 : i64
+  //   return %cst : i64
+  // }
+  //
+  // func.func private @callee.init_0(%arg0: !quake.veq<?>): !quake.veq<?> {
+  //   %cst = arith.constant 1.5707963267948966 : f64
+  //   %1 = quake.extract_ref %arg0[0] : (!quake.veq<2>) -> !quake.ref
+  //   quake.ry (%cst) %1 : (f64, !quake.ref) -> ()
+  //   return %arg0
+  // }
+  // ```
+  // clang-format on
+  if (simState->getKernelInfo().has_value()) {
+    auto [calleeName, calleeArgs] = simState->getKernelInfo().value();
+
+    std::string calleeKernelName =
+        cudaq::runtime::cudaqGenPrefixName + calleeName;
+
+    auto ctx = builder.getContext();
+    auto loc = builder.getUnknownLoc();
+
+    auto code = cudaq::get_quake_by_name(calleeName, /*throwException=*/false);
+    assert(!code.empty() && "Quake code not found for callee");
+    auto fromModule = parseSourceString<ModuleOp>(code, ctx);
+
+    static unsigned counter = 0;
+    std::string initName = calleeName + ".init_" + std::to_string(counter);
+    std::string initKernelName = cudaq::runtime::cudaqGenPrefixName + initName;
+
+    std::string numQubitsName =
+        calleeName + ".num_qubits_" + std::to_string(counter++);
+    std::string numQubitsKernelName =
+        cudaq::runtime::cudaqGenPrefixName + numQubitsName;
+
+    auto calleeFunc = fromModule->lookupSymbol<func::FuncOp>(calleeKernelName);
+    assert(calleeFunc && "callee is missing");
+
+    // Create `callee.init_N` and `callee.num_qubits_N` used for
+    // `quake.get_state` replacement later in ReplaceStateWithKernel pass
+    createInitFunc(builder, sourceMod, calleeFunc, initKernelName);
+    createNumQubitsFunc(builder, sourceMod, calleeFunc, numQubitsKernelName);
+
+    // Create substitutions for the `callee.init_N` and `callee.num_qubits_N`.
+    converter.genCallee(initName, calleeArgs);
+    converter.genCallee(numQubitsName, calleeArgs);
+
+    // Create a substitution for the state pointer.
+    auto statePtrTy =
+        cudaq::cc::PointerType::get(cudaq::cc::StateType::get(ctx));
+    return builder.create<quake::GetStateOp>(
+        loc, statePtrTy, builder.getStringAttr(numQubitsKernelName),
+        builder.getStringAttr(initKernelName));
+  }
+
+  TODO("cudaq::state* argument synthesis for quantum hardware for c functions");
```

Comment:
Why doesn't this use `cudaq::qkernel`? That's the vehicle for capturing quantum kernels in general-purpose C++ code contexts. Since `cudaq::get_state()` may appear anywhere, I don't understand how we know what kernel is being captured.

There is the additional issue of captured arguments as well...